src/libs/elektra/key.c

```diff
@@ -362,32 +362,54 @@ int f (const Key * source)
  * @code
 int g (const Key * source, KeySet * ks)
 {
-	Key * dup = keyDup (source);
+	Key * dup = keyDup (source, 0);
 	// work with duplicate
 	ksAppendKey (ks, dup);
 	// ksDel(ks) will also free the duplicate
 	// source remains unchanged.
 }
  * @endcode
  *
- * Duplication of keys should be preferred to keyNew(),
- * because data like owner can be filled with a copy
- * of the key instead of asking the environment.
- * It can also be optimized in the checks, because the keyname
- * is known to be valid.
+ * Depending on what @p flags are given, only parts of the key will be copied.
+ * The supported #elektraKeyFlags values are:
+ *   - #KEY_NAME   - copies the name
+ *   - #KEY_VALUE  - copies the value
+ *   - #KEY_META   - copies all meta keys
+ *   - #KEY_BINARY - creates a binary key, even if @p source is not binary
```

Comment:
Yes, it probably wasn't a good addition. The main reason was to define what key value you get, when the value is not copied (`NULL` or `""`). Maybe `KEY_NULL` is a better flag for this.

So if `KEY_NULL` is set and `KEY_VALUE` is not, you get a `NULL` value. If neither is set you get `""`. If `KEY_VALUE` is set we ignore `KEY_NULL` and just copy the value.

Reply:
Or we simply make the keys always `""` if the value does not get copied? Or whatever behavior we want of keyNew without any flags, I think this should be identical.

I think it will be confusing if some flags modify the copy behavior and other flags modify other behavior.

Reply:
We definitely shouldn't explicitly set `""`, that could lead to a lot use useless allocations.

Probably, both `Key * k = keyDup (orig, KEY_NAME)` and `Key * k = keyNew (name, KEY_END)` should set `k->data.c = NULL`. This doesn't cause an allocation and using `keyString (k)` it still appears as `""`.

Reply:
> We definitely shouldn't explicitly set "", that could lead to a lot use useless allocations.

I meant only from the user's perspective. If we really allocate is an implementation detail. (Like "/" as keyname.)

>  and using keyString (k) it still appears as ""

Exactly. And nobody is supposed to use `k->data.c` anyway.